src/proxy.rs

```diff
@@ -60,79 +60,63 @@ impl Proxy {
-                                },
-
-                                'r' => {
-                                    query.get_rules(&server, &mut writer);
-
-                                    let result = writer.into_inner();
-                                    match socket.send_to(&result, src) {
-                                        Ok(_) => (),
-                                        Err(e) => println!("Failed to send packet: {}", e)
-                                    }
-
-                                },
-
-                                'c' => {
-                                    query.get_players(&server, &mut writer);
-
-                                    let result = writer.into_inner();
-                                    match socket.send_to(&result, src) {
-                                        Ok(_) => (),
-                                        Err(e) => println!("Failed to send packet: {}", e)
-                                    }                                    
-                                },
-
-                                'd' => {
-                                    query.get_player_details(&server, &mut writer);                                
-
-                                    let result = writer.into_inner();
-                                    match socket.send_to(&result, src) {
-                                        Ok(_) => (),
-                                        Err(e) => println!("Failed to send packet: {}", e)
-                                    }                                       
-                                },
-
-                                'p' => {
-                                    query.get_ping(&mut writer);                                      
-
-                                    let result = writer.into_inner();
-                                    match socket.send_to(&result, src) {
-                                        Ok(_) => (),
-                                        Err(e) => println!("Failed to send packet: {}", e)
-                                    }                                    
-                                },
-
-                                _ => ()
-                            };
-                        }
-
-                        //println!("{:?}", packet);
-                    }
+                let src = match socket.recv_from(&mut buf) {
+                    Ok((_, src)) => src,
                     Err(e) => {
                         println!("couldn't recieve a datagram: {}", e);
+                        continue;
+                    }
+                };
+
+                let data: &[u8] = &buf;
+                let packet = Packet::new(&mut Cursor::new(data));
+
+                if packet.is_query_packet() {
+                    let mut writer = Cursor::new(vec![]);
+                    let query = Query::new(packet.clone());
+
+                    let send = match packet.data[0] as char {
```

Comment:
This would be a nicer interface if the match returned the packet to forward, and then the internals can decide whether or not to forward to the backend or use a cached copy. This pattern could then be abstracted further into a form of filter. This would then allow you to filter specific hacks/cheats/expoits using known patterns.

Reply:
I'm not good with patterns, so could you explain or show me the pattern I need to learn here?